Update(mnist):ps/worker cluster framework

DjangoPeng opened the pull request:
I updated the cluster setter function of `mnist_replica.py` with the latest and easy-understanding way in distributed tutorial [here](https://www.tensorflow.org/versions/r0.9/how_tos/distributed/index.html#putting-it-all-together-example-trainer-program).
1. `worker_grpc_url` -> ps_hosts / worker_hosts / server.target
2.Add the `FLAGS.job_name` to distinguish the `ps job` from `worker job`

DjangoPeng committed:
Modify the cluster setter function

DjangoPeng committed:
Revert "Modify the cluster setter function"

This reverts commit 83d7e56aa3ef1a8926663bbe2b4f177ce128af26.

DjangoPeng committed:
Update(mnist): fix the confusing cluster setter

tensorflow-jenkins commented:
Can one of the admins verify this patch?

rmlarsen commented:
@tensorflow-jenkins test this please

DjangoPeng commented:
@rmlarsen Thanks for your reply and test

rmlarsen commented:
@mrry please review this at your convenience.

DjangoPeng commented:
@rmlarsen Thank you !
@mrry Could you have a look at this request? I think it's a useful update

rmlarsen commented:
@mrry friendly ping :-)

mrry commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -56,9 +56,9 @@
                      "Total number of workers (must be >= 1)")
-flags.DEFINE_integer("num_parameter_servers", 2,
+flags.DEFINE_integer("num_parameter_servers", 1,
                      "Total number of parameter servers (must be >= 1)")
```
These flags are redundant if you have the two lists `--ps_hosts` and `--worker_hosts`. However, if you delete them, there will be consequences for the automated tests that run this code. @caisq: Can you comment on any other changes that would be necessary?

mrry commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -88,34 +94,28 @@
-  Args:
-    num_parameter_servers: Number of parameter servers
-    num_workers: Number of workers
+def get_cluster_setter():
```
This function is confusingly named (what is a "cluster setter"? why does it return a server and a cluster?). Since it's only called from one place, I'd inline it at the call site, and the code will be easier to follow.

mrry commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -88,34 +94,28 @@
-  for k in range(num_workers):
-    worker_spec.append("%s%d:%d" % (WORKER_PREFIX, k, FLAGS.grpc_port))
+  print(ps_spec)
+  print(worker_spec)
```
Delete these print statements.

mrry commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -137,128 +137,133 @@ def main(unused_argv):
+      if FLAGS.replicas_to_aggregate is None:
+        replicas_to_aggregate = FLAGS.num_workers
+      else:
+        replicas_to_aggregate = FLAGS.replicas_to_aggregate
```
Move this `if` statement to the same block where the `tf.train.SyncReplicasOptimizer` is constructed.

DjangoPeng commented:
@mrry You are right! I have updated all your comments:
- Delete the useless FLAGS arguments and info outputs
- Add the GPU allocation 
- Adjust the code structure
Please check the latest commit! Thanks for your comments.

DjangoPeng committed:
Update mnist_replica.py

Delete the useless FLAGS arguments and info outputs
Add the GPU allocation
Adjust the code structure Please check the latest commit! Thanks for your comments.

mrry commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -52,29 +52,24 @@
                      "the master worker task the performs the variable "
                      "initialization ")
 flags.DEFINE_integer("num_workers", 2,
                      "Total number of workers (must be >= 1)")
```
`num_workers` is also redundant with the `worker_hosts` flag.

mrry commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -52,29 +52,24 @@
-flags.DEFINE_integer("num_parameter_servers", 1,
-                     "Total number of parameter servers (must be >= 1)")
+flags.DEFINE_integer("num_gpu", 4,
+                     "Total number of gpus for each machine (must be >= 1")
```
Can you set the default here to be 1, and make it so that 0 is also a valid value? (You'll need to change the code for constructing the worker device, so that it uses `"/cpu:0"` when `FLAGS.num_gpus` is 0.)

Also, can you rename this flag from `num_gpu` to `num_gpus` for consistency with the other flags?

DjangoPeng replied:
But if there are 2 workers both use `"/cpu:0"`, it may cause the problem, like this:
```
Traceback (most recent call last):
  File "mnist_replica.py", line 256, in <module>
    tf.app.run()
  File "/Users/Patient/anaconda/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 30, in run
    sys.exit(main(sys.argv))
  File "mnist_replica.py", line 195, in main
    global_step=global_step)
  File "/Users/Patient/anaconda/lib/python2.7/site-packages/tensorflow/python/training/supervisor.py", line 327, in __init__
    self._verify_setup()
  File "/Users/Patient/anaconda/lib/python2.7/site-packages/tensorflow/python/training/supervisor.py", line 863, in _verify_setup
    "their device set: %s" % op)
ValueError: When using replicas, all Variables must have their device set: name: "global_step"
```
So I have to verify if there is only 1 ps and 1 worker to use the `"/cpu:0"` when they set `num_gpus = 0`. Or do I need to set a new flag named `num_cpus` to handle this situation when they have more than 1 workers but don't have gpus.

mrry commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -94,68 +89,38 @@
   
   if FLAGS.job_name == "ps":
     server.join()
   elif FLAGS.job_name == "worker":
```
Add an `else:` block to this statement that raises an exception if `FLAGS.job_name` is not set, or unknown.

DjangoPeng replied:
Sure thing, that will be more robust and security

mrry commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -183,13 +148,19 @@ def main(unused_argv):
+          replicas_to_aggregate = FLAGS.num_workers
+        else:
+          replicas_to_aggregate = FLAGS.replicas_to_aggregate
+
```
This `if FLAGS.sync_replicas:` block can be combined with the one below (on L158&ndash;164).

DjangoPeng committed:
Update mnist_replica.py

1. Delete the redundant argument 'FLAGS.num_workers' 
2. Change:'num_gpu' to 'num_gpus', and support to use'/cpu:0' instead of '/gpu:#' when setting 'num_gpus=0' 
3. Combine the 'if FLAGS.sync_replicas' statement to `SyncReplicasOptimizer` block.
4. Add `FLAGS.job_name`, 'FLAGS.task_index' and 'FLAGS.num_gpus' available check. It will raise an exception if any of them is not set, or unknown

DjangoPeng commented:
I have updated again, and the latest version fixed the bugs as below:
- Delete the redundant argument `FLAGS.num_workers` 
- `num_gpu` -> `num_gpus`
- Support to use `/cpu:0` instead of `/gpu:#` when setting `num_gpus=0` 
- Combine the `if FLAGS.sync_replicas` statement to `SyncReplicasOptimizer` block.
- Add `FLAGS.job_name`, `FLAGS.task_index` and `FLAGS.num_gpus` available check. That is, it will raise an exception if any of them is not set, or unknown

DjangoPeng commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -94,32 +93,48 @@ def main(unused_argv):
   if FLAGS.download_only:
     sys.exit(0)
 
+  if FLAGS.job_name is None or FLAGS.job_name == "":
```
Check if the `FLAGS.job_name` and `FLAGS.task_index` are available

DjangoPeng commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -94,32 +93,48 @@ def main(unused_argv):
 
+  # Get the number of workers 
+  num_workers = len(worker_spec)
+
```
Get the `num_workers` from `worker_hosts` and delete the redundant flags `num_workers`

DjangoPeng commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -94,32 +93,48 @@ def main(unused_argv):
+    elif FLAGS.num_gpus == 0:
+      # Just allocate the CPU to worker server
+      cpu = 0
+      worker_device = "/job:worker/task:%d/cpu:%d" % (FLAGS.task_index, cpu)
```
Support to use `/CPU:0`

DjangoPeng commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -94,32 +93,48 @@ def main(unused_argv):
-    gpu = (FLAGS.task_index % FLAGS.num_gpu)
+    if FLAGS.num_gpus > 0:
+      if FLAGS.num_gpus < num_workers:
+        raise ValueError("number of gpus is less than number of workers")
```
Check if `num_gpus` is equal to `num_workers` when `num_gpus>0`

DjangoPeng commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -148,18 +163,18 @@ def main(unused_argv):
       if FLAGS.sync_replicas:
         if FLAGS.replicas_to_aggregate is None:
-          replicas_to_aggregate = FLAGS.num_workers
+          replicas_to_aggregate = num_workers
```
Combine `if FLAGS.sync_replica` statement to `SyncReplicasOptimizer` block

DjangoPeng committed:
Fix a miss word in the comment

DjangoPeng commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -52,14 +52,13 @@
-                     "Total number of gpus for each machine (must be >= 1")
+flags.DEFINE_integer("num_gpus", 1,
+                     "Total number of gpus for each machine."
+                     "If you don't   GPU, please set it to '0'")
```
`num_gpu` -> `num_gpus`

DjangoPeng commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -52,33 +52,33 @@
-                     "Total number of parameter servers (must be >= 1)")
+flags.DEFINE_integer("num_gpus", 1,
+                     "Total number of gpus for each machine."
+                     "If you don't use GPU, please set it to '0'")
```
`num_gpu` -> `num_gpus`

DjangoPeng committed:
Fix a miss word in the comment

DjangoPeng committed:
Change FLAGS.sync_replicas default value to "True"

DjangoPeng commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -69,7 +69,7 @@
 flags.DEFINE_integer("batch_size", 100, "Training batch size")
 flags.DEFINE_float("learning_rate", 0.01, "Learning rate")
-flags.DEFINE_boolean("sync_replicas", False,
+flags.DEFINE_boolean("sync_replicas", True,
```
Set the flags arguments default value to `True`, because the program is a distributed training program.

DjangoPeng commented:
@mrry Please check the update~

DjangoPeng commented:
@mrry Please check the updates

mrry commented:
OK, these changes all look good to me. However, I think they'll cause the existing distributed tests to fail, because they assume that the old flags are valid. I'm not sure if the automated tests cover these, so I'll assign this to @caisq, who wrote the test and can confirm.

mrry commented:
@tensorflow-jenkins test this please.

DjangoPeng commented:
@mrry Thanks for your reply!

DjangoPeng commented:
@caisq Please Check and merge, it has passed all automatically checks

DjangoPeng commented:
@caisq @rmlarsen  @mrry  Please Check the updates ):

DjangoPeng commented:
@danmane Could you help me check and merge the PR? It has been a long time to wait for reviews, but no assignees come. Moreover, mrry has reviewed and checked the code before, and you can look through the  PR timeline~ :)

mrry commented:
@DjangoPeng Before we can submit this, please update the [distributed tests](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/dist_test) that use this script. In particular [`dist_mnist_test.sh`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/dist_test/scripts/dist_mnist_test.sh) makes assumptions about the flags that `mnist_replica.py` accepts. @caisq can advise on how to run this test.

DjangoPeng commented:
@mrry I'll update is today. It would be better If @caisq can give some suggestions

caisq commented on tensorflow/tools/dist_test/python/mnist_replica.py:
```diff
@@ -52,33 +52,33 @@
-                    "Worker GRPC URL (e.g., grpc://1.2.3.4:2222, or "
-                    "grpc://tf-worker0:2222)")
-flags.DEFINE_boolean("sync_replicas", False,
+flags.DEFINE_boolean("sync_replicas", True,
```
Can we default this to False for now? There is an unresolved issue in sync_replicas training in open-source TensorFlow.

DjangoPeng replied:
Set it back to `False` now

caisq commented:
LGTM except for the default value of sync_replicas, which should be False until the issue is resolved. We can merge this PR after @DjangoPeng fixes that. Then I will prepare a PR to update dist_mnist_test.sh accordingly myself.

Commit:
Add test for ps/worker and sync_replicas

DjangoPeng commented:
@caisq @mrry I have updated the `dist_mnist_test.sh` and it works well in `sync_replicas` mode. You can check and run the script.

Commit:
Set default  sync_replicas = False

DjangoPeng commented:
@tensorflow-jenkins test this please.

Commit:
Revert the change of 'sync-replicas` argument

DjangoPeng commented:
@caisq I have reverted the change of 'sync_replicas` argument in `dist_mnist_test.sh`. Now, it would exit and output ERROR INFO in case of `sync_replicas`

mrry commented:
Thanks @DjangoPeng and thanks @caisq for undertaking to fix the test (although hopefully less work will be necessary with the latest changes). This looks good to me. I'll run the tests and then we should be good to merge!

mrry commented:
@tensorflow-jenkins test this please.

DjangoPeng commented:
@mrry Thanks for your reply. I hope my update of `dist_mnist_test.sh` work well, if so, @caisq don't need to fix for the latest changes.

icklerly commented:
Hi @DjangoPeng ,
I have a question regarding the mnist_replica.py program. 
If I use only one machine (localhost) and set internally multiple workers everything runs fine. 
But if I want to use e.g. 2 worker running on different machines only the worker with index 0 runs through (does not wait for the other worker with index 1) and the worker with index 1 complains about global Variables not set.
I hope you can point me in the right direction!

DjangoPeng commented:
@icklery What TF version dou you used? Could you show me your parameters?
